pyccel/ast/internals.py

```diff
@@ -314,6 +324,52 @@ def is_temp(self):
         """
         return self._is_temp
 
+class AnnotatedPyccelSymbol(Basic):
+    """
+    Class representing a symbol in the code which has an annotation.
+
+    Symbolic placeholder for a Python variable, which has a name but no type yet.
+    This is very generic, and it can also represent a function or a module.
```

Comment:
This is what I did first. It caused problems in PR #1523. But I can't remember exactly what problems.
It makes it harder to detect if the `AnnotatedPyccelSymbol` is not correctly visited as `_visit_PyccelSymbol` is called instead.
I think there may also be a few places where it is useful to differentiate between `PyccelSymbol` and `AnnotatedPyccelSymbol`. As is this is simple, but if `AnnotatedPyccelSymbol` inherits from `PyccelSymbol` then the condition becomes more complex:
```python
if isinstance(var, PyccelSymbol) and not isinstance(var, AnnotatedPyccelSymbol)
```

If you think this is important then I can go back to the version where `AnnotatedPyccelSymbol` is a subclass.

I think I may have also had a problem in the scope as the symbols are used as keys the first time they are encountered. When there is a variable header the first time they are encountered they are `AnnotatedPyccelSymbol`s but then we look up the variable with a `PyccelSymbol`. This may have been the problem that caused me to change strategy, but thinking about it it should be possible to correct it by implementing `__eq__` instead.

Reply:
An additional argument comes from error raising. Errors are easily raised from an `AnnotatedPyccelSymbol`. This object should therefore have a `fst` property so the line and symbol can be shown in the error message. The handling of this is already done in `Basic`, but it would need to be added if `AnnotatedPyccelSymbol` subclassed `PyccelSymbol` as `PyccelSymbol`s are reused and are therefore not easily described by a single `fst` pattern.

Reply:
If I implement the duplicate methods then the tests seem to pass so I leave it up to you to decide which strategy you think makes more sense

Reply:
`PyccelSymbol` objects are immutable, but the `fst` property of `AnnotatedPyccelNode` needs to be set after creating an object of this class. Hence it makes sense that `AnnotatedPyccelNode` should subclass `Basic` as in the current implementation.